### DISCUSSION TITLE
No support for Docker compatibility mode enablement button for Linux (use documentation only)

### DISCUSSION THREAD

**Discussion by cdrage** [podman-desktop/podman-desktop#1922] [General]
Unfortunately, we wont be able to implement an "automatic" compatibility-mode button for Linux.

This is due to:
* Flatpak's sandbox model for no access to system services
* If we were to enable it manually with --socket=system-bus when building, it's only recommended for dev environments (see: https://docs.flatpak.org/_/downloads/en/latest/pdf/ and where it says: "Access to the entire bus with --socket=system-bus or --socket=session-bus should be avoided, unless the application is a development tool."
* Despite podman desktop being a development tool, opening up the system socket access just for compatibility enablement would be a big security risk.
* Too many libraries / upkeep for implementation in order to implement automatic enablement. Better to "tell" the developer to open a terminal and type in `systemctl enable --now podman.socket`.

I went ahead and implemented it here: https://github.com/containers/podman-desktop/pull/1847 unfortunately this will only work on non-flatpak environments. I only realized until after that Flatpak permissions are highly sandboxed for systemd / systemctl calls.

When trying to implement it within electron + flatpak, I created the code necessary to communicate to the API here:
https://gist.github.com/cdrage/beb190a32eaf1909bf19c76f93f1f926

Despite API calls, this unfortunately this requires us to add:
```
finish-args:
  - --socket=system-bus
```

To the electron-builder YAML which is not recommended by Flatpak for us to do.

This would also require us to add the `org.freedesktop.DBus GLib` runtime for enablement.

**Solution:**

Instead we should prompt the user to just type in `systemctl enable --now podman.socket` and have Podman Desktop detect that the socket has indeed been enabled / etc.

**Comment by benoitf** [comment]
we're able to call with flatpak some commands using flatpak-spawn. It does not work ?

**Reply by cdrage** [reply]
@benoitf Not well. I'm going to  modify it and find a way to do it without so many libraries / API calls.

**Reply by cdrage** [reply]
The problem is being able to run it under sudo for `flatpak-spawn` which is causing most of the issues, I've opened up the PR here if you were curious with regards to my development: https://github.com/containers/podman-desktop/pull/1925

Two issues:

1. Issue with sudo-prompt  on on non-flatpak environments: 

![Screenshot 2023-04-03 at 11 31 18 AM](https://user-images.githubusercontent.com/6422176/229557515-33981f10-0fde-4da6-b918-50acfae9a400.png)

Which can be fixed by applying https://github.com/jorangreef/sudo-prompt/pull/123

2. Running `flatpak-spawn --host sudo systemctl enable --now podman.socket`.

See: https://discussion.fedoraproject.org/t/host-spawn-a-better-way-to-run-host-commands-from-inside-your-toolbox/40876/3 and https://github.com/1player/host-spawn

We'd have to implement a way to successfully run sudo / permissions.


I'm going to try a few more things (in Fedora we can have it force prompt to ask for password natively) and test on Ubuntu as well.

**Comment by cdrage** [comment]
Closing in favour of: https://github.com/containers/podman-desktop/pull/1925

On Fedora / Ubuntu, they will automatically initiate a prompt to enter a password for privileges when using flatpak-spawn, but not on Debian.
